### PR TITLE
ASoC: SOF: Intel: add dummy echo reference BE link 

### PR DESCRIPTION
**Pull request by bardliao**
DPAM needs a BE to connect to a FE. The echo reference audio stream
will be the speaker playback stream. The dummy BE link will be
associated with a dummy SOF widget (swidget) which is invisible to
SOF.

### CODE REVIEW

File: sound/soc/intel/boards/sof_sdw.c [hunk 1]
```diff
@@ -1443,6 +1456,10 @@ static int sof_card_dai_links_create(struct device *dev,
 		init_dai_link(dev, links + link_index, be_id, name, 1, 1,
 				cpus + cpu_id, 1, ssp_components, 1, NULL, NULL);
 	}
+	/* Echo reference */
```
Comment (RanderWang) [inline sound/soc/intel/boards/sof_sdw.c:1459]: do we need to increase link_index and be_id before Echo reference ?

**Comment by RanderWang** [comment]
@bardliao  do you have topology link ? it is much easy to understand it with topology

**Comment by bardliao** [comment]
> @bardliao do you have topology link ? it is much easy to understand it with topology

Yes, thanks @RanderWang. I didn't notice that there is no `INC_ID(be_id, cpu_id, link_index);` in `if (sof_sdw_quirk & SOF_SSP_BT_OFFLOAD_PRESENT) {}`. Added it now.

**Comment by bardliao** [comment]
cavs-sdw-echo-reference

Please see the diagram for reference. The copier.ALH.16.1 -> copier.host.15.1 is the echo reference pipeline. And the dummy BE link is for copier.ALH.16.1. Also, copier.ALH.16.1 and the "copier.ALH.16.1 -> copier.host.15.1" route are invisible by the firmware.

File: sound/soc/intel/boards/sof_sdw.c [hunk 2]
```diff
@@ -1442,7 +1455,12 @@ static int sof_card_dai_links_create(struct device *dev,
 		cpus[cpu_id].dai_name = cpu_name;
 		init_dai_link(dev, links + link_index, be_id, name, 1, 1,
 				cpus + cpu_id, 1, ssp_components, 1, NULL, NULL);
+		INC_ID(be_id, cpu_id, link_index);
 	}
+	/* Echo reference */
+	init_dai_link(dev, links + link_index, be_id, "Echo-reference", 0, 1,
+			sof_dummy_component, 1, dummy_component, 1, NULL, NULL);
+	num_links++;
```
Comment (ujfalusi) [inline sound/soc/intel/boards/sof_sdw.c]: Should this echo reference be optional?
Like extending the `mach_params` with a bool.

File: sound/soc/intel/boards/sof_sdw.c [hunk 3]
```diff
@@ -1442,7 +1455,12 @@ static int sof_card_dai_links_create(struct device *dev,
 		cpus[cpu_id].dai_name = cpu_name;
 		init_dai_link(dev, links + link_index, be_id, name, 1, 1,
 				cpus + cpu_id, 1, ssp_components, 1, NULL, NULL);
+		INC_ID(be_id, cpu_id, link_index);
 	}
+	/* Echo reference */
+	init_dai_link(dev, links + link_index, be_id, "Echo-reference", 0, 1,
+			sof_dummy_component, 1, dummy_component, 1, NULL, NULL);
```
Comment (ujfalusi) [inline sound/soc/intel/boards/sof_sdw.c:1461]: links were allocated at line 1248 / 1261 to hold `num_links` number of links, here you will be over that.

**Review by ujfalusi — COMMENTED** [review]
@bardliao, I'm not saying that I understand the mechanics behind this. How a full dummy link is related to Echo-reference?

File: sound/soc/sof/intel/hda-dai.c [hunk 4]
```diff
@@ -916,6 +916,17 @@ struct snd_soc_dai_driver skl_dai[] = {
 	},
 },
 #endif
+{
+	.name = "SOF Dummy",
+	.playback = {
+		.channels_min = 1,
+		.channels_max = 8,
+	},
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-dai.c:924]: if this is an echo reference is this not capture-only?

File: sound/soc/sof/intel/hda-dai.c [hunk 5]
```diff
@@ -916,6 +916,17 @@ struct snd_soc_dai_driver skl_dai[] = {
 	},
 },
 #endif
+{
+	.name = "SOF Dummy",
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-dai.c:920]: SOF echo reference?

File: sound/soc/sof/intel/hda-common-ops.c [hunk 6]
```diff
@@ -80,7 +80,7 @@ struct snd_sof_dsp_ops sof_hda_common_ops = {
 
 	/* DAI drivers */
 	.drv		= skl_dai,
-	.num_drv	= SOF_SKL_NUM_DAIS,
+	.num_drv	= SOF_SKL_NUM_DAIS + 1, /* + 1 for dummy DAI */
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-common-ops.c:83]: echo reference DAI?

File: sound/soc/intel/boards/sof_sdw.c [hunk 7]
```diff
@@ -404,6 +404,19 @@ static struct snd_soc_dai_link_component dmic_component[] = {
...
+		.dai_name = "snd-soc-dummy-dai",
+	}
+};
+
+static struct snd_soc_dai_link_component sof_dummy_component[] = {
+	{
+		.dai_name = "SOF Dummy",
+	}
+};
+
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw.c:419]: I am lost on why we need this capability for SoundWire devices. The echo reference is typically provided by amplifiers, and we're not planning any topology change to add this echo reference in firmware, are we?

What are we trying to enable here?

Comment (plbossart) [reply sound/soc/intel/boards/sof_sdw.c:419]: And actually in the case of SoundWire, it's rather easy to enable an echo reference by reading the same bits from the bus, as we did for the sdw-mockup. We don't need any firmware or virtual dai, we could use an existing dai to create the echo reference. The problem would be to provide the same parameters as the speaker dai.

File: sound/soc/intel/boards/sof_sdw.c (hunk 3 above)
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c:1461]: > links were allocated at line 1248 / 1261 to hold `num_links` number of links, here you will be over that.

Thanks @ujfalusi. Indeed, I should increase  `num_links` before that. I don't know how did it work on my test.

File: sound/soc/intel/boards/sof_sdw.c (hunk 2 above)
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c]: > Should this echo reference be optional? Like extending the `mach_params` with a bool.

I thought about this. But I think we can always create an echo reference BE dailink even we don't use it. @plbossart What do you think?

File: sound/soc/sof/intel/hda-dai.c (hunk 5 above)
Comment (bardliao) [reply sound/soc/sof/intel/hda-dai.c:920]: > SOF echo reference?

@plbossart I thought about this when I added the new cpu dai. The dai will be for echo reference only if I called it as "echo reference" and set capture capability only. But if we call it a dummy dai and set both playback and capture capability, we can use it for any cases that need a dummy cpu dai.
And we can use the dailink name to describe the purpose of the BE link. What do you think?

File: sound/soc/intel/boards/sof_sdw.c (hunk 7 above)
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c:419]: > I am lost on why we need this capability for SoundWire devices. The echo reference is typically provided by amplifiers, and we're not planning any topology change to add this echo reference in firmware, are we?
> 
> What are we trying to enable here?

@plbossart I am trying to provide an echo reference solution that the amplifiers don't provide echo reference.
And yes, I am working on a topology change for echo reference. Please see https://github.com/thesofproject/linux/pull/3995#issuecomment-1305366527 for reference.
Indeed, we can read the same bits from the bus on SoundWire. However, adding an echo reference route in topology can be a general solution for all interfaces.

Comment (plbossart) [reply sound/soc/intel/boards/sof_sdw.c:419]: Sorry @bardliao I am losing the plot and I don't recall who's asking for this capability? I know you've explained this already but it's been weeks now and I have lost all context.

All SoundWire amplifiers I am aware of have an echo reference, we just chose not to support it.
We could use an SSP loopback for all amplifiers that don't provide any feedback.
Why would we add a 'generic' solution when there are simpler solutions?

**Comment by bardliao** [comment]
There is no requirement. Close it

**Pull request closed without merging**